File: server/internal/accounts/simplefin.go
package accounts

import (
	"context"
	"fmt"
	"log/slog"

	"tallyo/internal/apierror"
	"tallyo/internal/clients"
	"tallyo/internal/graph/model"

	"github.com/samber/lo"
)

// SimpleFinLinkStore is the subset of Store methods required by SimpleFinService.
type SimpleFinLinkStore interface {
	OwnerByID(ctx context.Context, id int64) (*model.Owner, error)
	CreateSimpleFinAccessToken(ctx context.Context, accessURL string, ownerID int64, label string) (*model.SimpleFinAccessToken, error)
	SimpleFinAccessTokenByID(ctx context.Context, id int64) (*model.SimpleFinAccessToken, error)
	LinkSimpleFinConnection(ctx context.Context, conn UpsertSimpleFinConnectionParams) (simpleFinConnID int64, connectionID int64, err error)
	SimpleFinConnectionsByTokenIDs(ctx context.Context, accessTokenIDs []int64) (map[int64][]*model.SimpleFinConnection, error)
	UpsertAccountWithExternalID(ctx context.Context, externalID string, account *model.Account) (int64, error)
	ResetSimpleFinTokenSyncedAt(ctx context.Context, id int64) error
	DeleteSimpleFinAccessToken(ctx context.Context, id int64) error
}

type SimpleFinService struct {
	Store  SimpleFinLinkStore
	Client clients.SimpleFinClient
	Events Publisher
	Log    *slog.Logger
}

// CreateAccessToken claims a setup token, persists the access URL, fetches
// connections/accounts, and publishes AccountsCreated events.
func (s *SimpleFinService) CreateAccessToken(
	ctx context.Context,
	setupToken string,
	ownerID int64,
	label string,
) (*model.CreateSimpleFinAccessTokenPayload, error) {
	owner, err := s.ownerByID(ctx, ownerID)
	if err != nil {
		return nil, err
	}

	accessURL, err := s.Client.Claim(ctx, setupToken)
	if err != nil {
		return nil, fmt.Errorf("claim setup token: %w", err)
	}

	token, err := s.Store.CreateSimpleFinAccessToken(ctx, accessURL, ownerID, label)
	if err != nil {
		return nil, fmt.Errorf("persist access token: %w", err)
	}
	token.Owner = owner

	tokenID := token.ID.Int64()

	accountSet, err := s.Client.GetAccounts(ctx, accessURL, clients.GetAccountsOpts{Pending: true})
	if err != nil {
		s.Log.Warn("initial simplefin fetch failed; token saved but no accounts yet", "error", err)
		return &model.CreateSimpleFinAccessTokenPayload{
			AccessToken: token,
			Connections: []*model.SimpleFinConnection{},
			Accounts:    []*model.Account{},
		}, nil
	}

	connectionIDByExternalConnID := make(map[string]int64, len(accountSet.Connections))
	connectionIDBySimpleFinConnID := make(map[int64]int64, len(accountSet.Connections))

	for _, sfConn := range accountSet.Connections {
		params := NewUpsertSimpleFinConnectionParams(tokenID, ownerID, sfConn)
		simpleFinConnID, connectionID, err := s.Store.LinkSimpleFinConnection(ctx, params)
		if err != nil {
			return nil, fmt.Errorf("link simplefin connection %s: %w", sfConn.ConnID, err)
		}
		connectionIDByExternalConnID[sfConn.ConnID] = connectionID
		connectionIDBySimpleFinConnID[simpleFinConnID] = connectionID
	}

	for _, sfAccount := range accountSet.Accounts {
		connectionID, ok := connectionIDByExternalConnID[sfAccount.ConnID]
		if !ok {
			s.Log.Warn("skipping simplefin account: connection not found", "conn_id", sfAccount.ConnID)
			continue
		}

		fields := FieldsFromSimpleFinAccount(sfAccount)
		account := &model.Account{
			Connection:  &model.Connection{ID: model.New(model.GlobalIDConnection, connectionID)},
			Owner:       owner,
			Name:        fields.Name,
			Type:        fields.Type,
			Mask:        fields.Mask,
			NeedsReview: fields.NeedsReview,
			Manual:      false,
		}
		if _, err := s.Store.UpsertAccountWithExternalID(ctx, fields.ID, account); err != nil {
			return nil, fmt.Errorf("upsert account %s: %w", fields.ID, err)
		}
	}

	connectionsByToken, err := s.Store.SimpleFinConnectionsByTokenIDs(ctx, []int64{tokenID})
	if err != nil {
		return nil, fmt.Errorf("fetch simplefin connections: %w", err)
	}
	connections := connectionsByToken[tokenID]
	if len(connections) > 0 {
		conn := connections[0]
		s.Events.Publish(AccountsCreated{
			ConnectionID: connectionIDBySimpleFinConnID[conn.ID.Int64()],
			Provider:     SourceTableSimpleFinConnection,
			SourceID:     conn.ID.Int64(),
		})
	}
	toAccounts := func(conn *model.SimpleFinConnection, _ int) []*model.Account { return conn.Accounts }
	allAccounts := lo.FlatMap(connections, toAccounts)

	return &model.CreateSimpleFinAccessTokenPayload{
		AccessToken: token,
		Connections: connections,
		Accounts:    allAccounts,
	}, nil
}

// DeleteAccessToken removes the access token and all dependent data.
func (s *SimpleFinService) DeleteAccessToken(ctx context.Context, id int64) error {
	return s.Store.DeleteSimpleFinAccessToken(ctx, id)
}

// ResetSync clears last_synced_at to force a full re-pull on the next tick.
func (s *SimpleFinService) ResetSync(ctx context.Context, id int64) (*model.SimpleFinAccessToken, error) {
	if err := s.Store.ResetSimpleFinTokenSyncedAt(ctx, id); err != nil {
		return nil, fmt.Errorf("reset simplefin sync: %w", err)
	}
	return s.Store.SimpleFinAccessTokenByID(ctx, id)
}

func (s *SimpleFinService) ownerByID(ctx context.Context, ownerID int64) (*model.Owner, error) {
	owner, err := s.Store.OwnerByID(ctx, ownerID)
	if err != nil {
		return nil, fmt.Errorf("lookup owner: %w", err)
	}
	if owner == nil {
		return nil, apierror.Publicf("invalid owner id %d", ownerID)
	}
	return owner, nil
}

File: server-rs/src/accounts/simplefin.rs
use std::collections::HashMap;

use anyhow::Result;
use sqlx::SqlitePool;

use crate::{
    accounts::{
        AccountsCreated, CreateSimpleFinAccessTokenPayload, EventBus, SimpleFinAccessToken, UpsertAccount,
        simplefin_types::{fields_from_simplefin_account, new_upsert_simple_fin_connection_params},
    },
    apierror::ApiError,
    clients::simplefin::{GetAccountsOpts, SimpleFinClient},
};

use super::store::{
    create_simple_fin_access_token, delete_simple_fin_access_token, owner_by_id, reset_simple_fin_token_synced_at,
    simple_fin_access_token_by_id, simple_fin_connections_by_token_ids, upsert_account,
};

pub struct SimpleFinService {
    pub pool: SqlitePool,
    pub client: SimpleFinClient,
    pub events: EventBus,
}

impl SimpleFinService {
    pub async fn create_access_token(
        &self,
        setup_token: &str,
        owner_id: i64,
        label: &str,
    ) -> Result<CreateSimpleFinAccessTokenPayload> {
        let owner = owner_by_id(&self.pool, owner_id)
            .await?
            .ok_or_else(|| ApiError::bad_input(format!("invalid owner id {owner_id}")))?;
        let access_url = self.client.claim(setup_token).await?;
        let token = create_simple_fin_access_token(&self.pool, &access_url, owner_id, label).await?;
        let account_set = match self
            .client
            .get_accounts(
                &access_url,
                GetAccountsOpts {
                    pending: true,
                    ..Default::default()
                },
            )
            .await
        {
            Ok(account_set) => account_set,
            Err(error) => {
                tracing::warn!(%error, "initial simplefin fetch failed; token saved but no accounts yet");
                return Ok(CreateSimpleFinAccessTokenPayload {
                    access_token: token,
                    connections: Vec::new(),
                    accounts: Vec::new(),
                });
            }
        };
        let mut connections_by_external_id = HashMap::with_capacity(account_set.connections.len());
        for connection in &account_set.connections {
            let params = new_upsert_simple_fin_connection_params(token.id, owner_id, connection);
            let (simple_fin_connection_id, connection_id) =
                super::store::link_simple_fin_connection(&self.pool, &params).await?;
            connections_by_external_id.insert(connection.conn_id.clone(), (simple_fin_connection_id, connection_id));
        }
        for simple_fin_account in &account_set.accounts {
            let Some((_, connection_id)) = connections_by_external_id.get(&simple_fin_account.conn_id) else {
                tracing::warn!(
                    conn_id = simple_fin_account.conn_id,
                    "skipping simplefin account: connection not found"
                );
                continue;
            };
            let fields = fields_from_simplefin_account(simple_fin_account);
            upsert_account(
                &self.pool,
                &UpsertAccount {
                    external_id: fields.id,
                    connection_id: Some(*connection_id),
                    owner_id: owner.id,
                    name: fields.name,
                    account_type: fields.account_type,
                    subtype: None,
                    mask: fields.mask,
                    notes: None,
                    closed: false,
                    hidden: false,
                    needs_review: fields.needs_review,
                },
            )
            .await?;
        }
        let connections = simple_fin_connections_by_token_ids(&self.pool, &[token.id]).await?;
        let connections = connections.get(&token.id).cloned().unwrap_or_default();
        if let Some(connection) = connections.first() {
            self.events.publish(AccountsCreated {
                connection_id: connection.connection_id,
                provider: crate::accounts::SourceTable::SimpleFinConnections,
                source_id: connection.connection.id,
            });
        }
        let connections = connections
            .into_iter()
            .map(|record| record.connection)
            .collect::<Vec<_>>();
        let accounts = connections
            .iter()
            .flat_map(|connection| connection.accounts.clone())
            .collect();
        Ok(CreateSimpleFinAccessTokenPayload {
            access_token: token,
            connections,
            accounts,
        })
    }

    pub async fn delete_access_token(&self, id: i64) -> Result<()> {
        delete_simple_fin_access_token(&self.pool, id).await
    }

    pub async fn reset_sync(&self, id: i64) -> Result<Option<SimpleFinAccessToken>> {
        reset_simple_fin_token_synced_at(&self.pool, id).await?;
        simple_fin_access_token_by_id(&self.pool, id).await
    }
}

#[cfg(test)]
mod tests {

    use anyhow::Result;
    use serde_json::json;
    use wiremock::MockServer;

    use crate::{
        accounts::{EventBus, SimpleFinService},
        clients::simplefin::SimpleFinClient,
        database::dbtest,
        testutil::{simplefin_mock, store::create_owner},
    };

    #[tokio::test]
    async fn claims_a_token_and_links_simplefin_accounts() -> Result<()> {
        let server = MockServer::start().await;
        let access_url = simplefin_mock::access_url(&server);
        simplefin_mock::mount_claim(&server, &access_url).await;
        simplefin_mock::mount_accounts(
            &server,
            json!({
                "connections":[{"conn_id":"bank","name":"Bank","org_url":"https://bank.example"}],
                "accounts":[
                    {"id":"unknown","conn_id":"bank","name":"Mystery"},
                    {"id":"checking","conn_id":"bank","name":"Joint Checking"}
                ]
            }),
        )
        .await;
        let pool = dbtest::open().await?;
        let owner = create_owner(&pool, "alex").await?;
        let events = EventBus::default();
        let mut subscriber = events.register_subscriber("test");
        let service = SimpleFinService {
            pool,
            client: SimpleFinClient::new()?,
            events,
        };
        let payload = service
            .create_access_token(&simplefin_mock::setup_token(&server), owner.id, "Bridge")
            .await?;
        assert_eq!(payload.connections.len(), 1);
        assert_eq!(payload.accounts.len(), 2);
        assert!(
            payload
                .accounts
                .iter()
                .any(|account| account.name == "Mystery" && account.needs_review)
        );
        assert_eq!(
            subscriber.recv().await.unwrap().provider,
            crate::accounts::SourceTable::SimpleFinConnections
        );
        service.reset_sync(payload.access_token.id).await?;
        service.delete_access_token(payload.access_token.id).await?;
        Ok(())
    }
}
